virttest/qemu_vm.py

```diff
@@ -4864,11 +4864,9 @@ def is_json_data(data):
                             v = parse_json(v)
                         if k != key:
                             continue
-                        if v != value:
-                            matched = False
-                        break
-                    else:
-                        matched = False
+                        if value in str(v):
```

Comment:
value can be a dict , so we cannot simply check it in this way, even if value is always a string, we may also cannot check it in this way, take block_auto_detect_size_increased as an example:

p_dict = {'file': '/home/kvm_autotest_root/images/stg.luks'}
but the item in block here is 
{'file': {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'}, so
k = 'file'
v =  {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'}
of course here value('/home/kvm_autotest_root/images/stg.luks) is in str(v), but its key('file') is not 'filename'. 
For luks image, it's reasonable to set p_dict = {'filename':'  '/home/kvm_autotest_root/images/stg.luks'}, but in this way, k != key, so we need to work out this. Some thoughts:
1. p_dict is a filter, we use this filter to get the exact node-name
2. p_dict can be a sub-set of value with essential information, here we don't need to set p_dict =  {'file':{'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'} }, only set p_dict = {'filename': '/home/kvm_autotest_root/images/stg.luks'} can get the exact device node. Give another example to represent a dedicated iscsi image: {"portal": "xxx", "target":"xxx", "lun":"0", "driver": "iscsi"}, we don't care about others.
3. do we need to pass p_dict as a dict with nested dict obj? e.g. {xxx{}}
@YongxueHong @nickzhq

Reply:
> value can be a dict , so we cannot simply check it in this way, even if value is always a string, we may also cannot check it in this way, take block_auto_detect_size_increased as an example:
> 
> p_dict = {'file': '/home/kvm_autotest_root/images/stg.luks'} but the item in block here is {'file': {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'}, so k = 'file' v = {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'} of course here value('/home/kvm_autotest_root/images/stg.luks) is in str(v), but its key('file') is not 'filename'. For luks image, it's reasonable to set p_dict = {'filename':' '/home/kvm_autotest_root/images/stg.luks'}, but in this way, k != key, so we need to work out this. Some thoughts:
> 
>     1. p_dict is a filter, we use this filter to get the exact node-name
> 
>     2. p_dict can be a sub-set of value with essential information, here we don't need to set p_dict =  {'file':{'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'} }, only set p_dict = {'filename': '/home/kvm_autotest_root/images/stg.luks'} can get the exact device node. Give another example to represent a dedicated iscsi image: {"portal": "xxx", "target":"xxx", "lun":"0", "driver": "iscsi"}, we don't care about others.
> 
>     3. do we need to pass p_dict as a dict with nested dict obj? e.g. {xxx{}}
>        @YongxueHong @nickzhq

Hi @nickzhq 
I remembered that I fixed to support luks format issue by https://github.com/avocado-framework/avocado-vt/pull/2629 .

Hi, @zhencliu 
I agree with your solution by fixing it from the caller(test case level)
suggestion:
```
        if data_image.image_format == 'luks':
            filename = get_image_json(img_tag, params, root_dir)
        device_name = vm.get_block({"filename": filename})
```
Thanks.

Reply:
Hi @YongxueHong , make sense, we can fix it in test case.

Another concern here, if we want to find an image with backing file, e.g. 
json:{"backing": {"driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/data.qcow2"}}, "driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}}  >>> J1

but the get_image_json will not return the json string with the backing info, it just return the following:
json:{"file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}, "driver": "qcow2"}    >>> J2

As you can see, if we translate the json string to the dict, J2 is a subset of J1.
So line 4867, we may want to check that value is a subset of v, if both value and v are dicts.
all([v[k] == value.get(k) for k in value])

Reply:
> Hi @YongxueHong , make sense, we can fix it in test case.
> 
> Another concern here, if we want to find an image with backing file, e.g. json:{"backing": {"driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/data.qcow2"}}, "driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}} >>> J1
> 
> but the get_image_json will not return the json string with the backing info, it just return the following: json:{"file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}, "driver": "qcow2"} >>> J2
> 


Yes, your concern makes sense. Anyway, from my understanding,  we could consider whether adjust the value of returning of the API `get_image_json` if we need this requirement after researching.
Thanks.


> As you can see, if we translate the json string to the dict, J2 is a subset of J1. So line 4867, we may want to check that value is a subset of v, if both value and v are dicts. all([v[k] == value.get(k) for k in value])